udpa/type/v1/typed_struct.proto

```diff
@@ -0,0 +1,38 @@
+syntax = "proto3";
+
+package udpa.type.v1;
+
+option java_outer_classname = "TypedStructProto";
+option java_multiple_files = true;
+option java_package = "com.github.udpa.udpa.type.v1";
+
+import "validate/validate.proto";
+import "google/protobuf/struct.proto";
+
+// A TypedStruct contains an arbitrary JSON serialized protocol buffer message with a URL that
+// describes the type of the serialized message. This is very similar to google.protobuf.Any,
+// instead of having protocol buffer binary, this employs google.protobuf.Struct as value.
+//
+// This message is intended to be embedded inside Any, so it shouldn't be directly referred
+// from other UDPA messages.
+//
+// When packing an opaque extension config, packing the expected type into Any is preferred
+// wherever possible for its efficiency. TypedStruct should be used only if a proto descriptor
+// is not available, for example if:
+// - A control plane sends opaque message that is originally from external source in human readable
+//   format such as JSON or YAML.
+// - The control plane doesn't have the knowledge of the protocol buffer schema hence it cannot
```

Comment:
It's not just the control plane -- even if the management server knows the protobuf schema, it may be that the DPLB won't know the protobuf schema.  For example, in gRPC, we have LB plugins that take their configs in JSON form, which means that gRPC will need to convert the config message for the LB plugin (see https://github.com/envoyproxy/envoy/pull/7744) to JSON, and it can't do that with an `Any` field, because it doesn't know the protobuf schema.

We have previously discussed the idea of adding a client capability that tells the management server that the client requires the content of all `Any` fields to be populated with a `Struct` instead of with an arbitrary protobuf; see https://github.com/envoyproxy/envoy/issues/4475#issuecomment-507439716.  My understanding from Harvey is that we will be working on the client capability mechanism as part of the xDS v3 API.

I assume the idea here is to use `TypedStruct` instead of `Struct` inside of the `Any`, which seems fine to me.  Conveying the type along with the `Struct` seems like useful information; in some cases, I could imagine that allowing clients to opportunistically convert a `TypedStruct` to the right protobuf schema if they happen to know it.

Reply:
> It's not just the control plane -- even if the management server knows the protobuf schema, it may be that the DPLB won't know the protobuf schema. For example, in gRPC, we have LB plugins that take their configs in JSON form, which means that gRPC will need to convert the config message for the LB plugin (see envoyproxy/envoy#7744) to JSON, and it can't do that with an Any field, because it doesn't know the protobuf schema.

I think this is a separate case that "A plugin expects google.protobuf.Struct (i.e. JSON)" type and doesn't have to have protobuf schema, no? That's the case Struct can still be used in Any directly without involving TypedStruct. And DPLB is only responsible to convert the Struct-in-Any to JSON format that the plugin expects.

Reply:
It depends whether the gRPC plugin uses protobuf or not, if the plugin expect JSON natively (i.e. it doesn't have schema at the first place) then it doesn't make a lot of sense to include a type_url. Though I'm fine either way if you think this is better for gRPC plugin.

In Envoy side I think we will hit similar issue with non-linked plugin, i.e. Envoy it self doesn't know the protobuf schema (i.e. descriptors are not linked) but the plugin knows, when we start deploying WASM extensions. Happy to discuss further on those use cases too. I'll add a TODO here.

Reply:
gRPC's current API for providing plugins with configuration does use JSON rather than protobuf, but that's really just an implementation detail of the way we've chosen to architect things; there could be other implementations in the future that could use protobuf instead.  Also, even though the plugins in our current implementation do use JSON internally, many config distribution mechanisms (xDS is only one of many that we support) are designed to work with protos, so many of our plugins will have a config proto defined even though the plugin itself doesn't use it.

I agree that this is likely to be a problem with Envoy plugins as well.  That probably strengthens the case for trying to find a general-purpose solution here.

Reply:
@lizan could you add an additional bullet point explaining that a DPLB that is only capable of understanding `Struct` should also use this? I think this addresses @markdroth's concern. It seems less confusing to have a single JSON embedding, and making the type URL explicit generally shouldn't hurt anyone.

Reply:
I added a TODO below already, not very thrilled to have TypedStruct where the underlying config is not typed (protobuf defined), having Any like:
```
type_url: udpa.type.v1.TypedStruct {
  type_url: google.protobuf.Struct {
    value: {(actual config)}
  }
}
```
for that case doesn't make sense with extra wrapping. That's why I said this is the case that "A plugin expects google.protobuf.Struct (i.e. JSON)", which we do already handle in Envoy: https://github.com/envoyproxy/envoy/blob/master/source/common/config/utility.cc#L274.

Reply:
I think the point is that if we make something like:
```
type_url: udpa.type.v1.TypedStruct {
  type_url: <the actual type URL> {
    value: {(actual config)}
  }
}
```
the standard way to embed in `Any`, it both solves the problem that gRPC has (it just wants a `Struct Value`, it doesn't care how it gets there) and also the other use cases you describe.

Reply:
My point is in some case `google.protobuf.Struct` is a legit type in top-level Any, when it is not typed. Anyway, either works, added a bullet.